File: Javascript/Linked List/M-19. Remove Nth Node From End of List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  const nodeMap = new Map<number, ListNode>();
  let node = head;
  let i = 0;
  while (node) {
    i++;
    nodeMap.set(i, node);
    node = node.next;
  }
  const prevTargetNodeIndex = i - n; // 2, 2
  const prevTargetNode = nodeMap.get(prevTargetNodeIndex);
  if (prevTargetNode) {
    if (prevTargetNode.next) prevTargetNode.next = prevTargetNode.next.next;
    else prevTargetNode.next = null;
    return head;
  } else {
    return head?.next || null;
  }
}

File: Linked List/M-19. Remove Nth Node From End of List/M-19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        node_map = {}
        ttl_nodes = 0
        node = head

        while node:
            ttl_nodes += 1
            node_map[ttl_nodes] = node
            node = node.next

        prev_target_node_idx = ttl_nodes - n
        prev_target_node = node_map.get(prev_target_node_idx)
        if prev_target_node:
            prev_target_node.next = prev_target_node.next.next
            return head
        else:
            return head.next
